File: sdk/python/errly.py
"""
Errly Python SDK — lightweight error monitoring

Usage:
    from errly import Errly

    errly = Errly(
        url="http://localhost:5080",
        api_key="your-api-key",
        project="my-service",
        environment="production",
        release="v1.2.3",
    )

    # Capture an exception
    try:
        risky_operation()
    except Exception as e:
        errly.capture_exception(e)

    # Capture a message
    errly.capture_message("Something weird happened", level="warning")

    # FastAPI / Flask middleware
    app = errly.instrument_fastapi(app)
"""

from __future__ import annotations

import json
import queue
import sys
import threading
import time
import traceback
import uuid
from datetime import datetime, timezone
from typing import Any, Optional
from urllib import request as urllib_request
from urllib.error import URLError


class Errly:
    def __init__(
        self,
        url: str,
        api_key: str = "",
        project: str = "",
        environment: str = "production",
        release: str = "",
        max_queue: int = 512,
    ):
        self.url = url.rstrip("/")
        self.api_key = api_key
        self.project = project
        self.environment = environment
        self.release = release

        self._queue: queue.Queue = queue.Queue(maxsize=max_queue)
        self._user: Optional[dict] = None
        self._breadcrumbs: list[dict] = []
        self._lock = threading.RLock()
        self._stop = threading.Event()

        # Background worker thread — daemon so it never blocks interpreter exit,
        # but flush() can be called explicitly to drain before shutdown.
        self._worker = threading.Thread(target=self._run, daemon=True)
        self._worker.start()

    # ── Public API ────────────────────────────────────────────────────────────

    def capture_exception(
        self,
        exc: Optional[BaseException] = None,
        extra: Optional[dict] = None,
    ) -> str:
        """Capture an exception with its stack trace."""
        if exc is None:
            exc_type, exc, tb = sys.exc_info()
        else:
            exc_type = type(exc)
            tb = exc.__traceback__

        event = self._new_event("error")
        event["exception"] = {
            "type": exc_type.__name__ if exc_type else "Exception",
            "value": str(exc),
            "module": getattr(exc_type, "__module__", ""),
        }
        event["stacktrace"] = _parse_traceback(tb)
        if extra:
            event["extra"] = extra

        self._enqueue(event)
        return event["id"]

    def capture_message(
        self,
        message: str,
        level: str = "info",
        extra: Optional[dict] = None,
    ) -> str:
        """Capture a plain message."""
        event = self._new_event(level)
        event["message"] = message
        if extra:
            event["extra"] = extra
        self._enqueue(event)
        return event["id"]

    def set_user(self, user_id: str = "", email: str = "", username: str = "") -> None:
        """Set the current user for all subsequent events."""
        with self._lock:
            self._user = {"id": user_id, "email": email, "username": username}

    def add_breadcrumb(
        self,
        message: str,
        category: str = "default",
        level: str = "info",
        crumb_type: str = "default",
    ) -> None:
        """Add a breadcrumb to the trail (max 50 kept)."""
        with self._lock:
            self._breadcrumbs.append({
                "timestamp": datetime.now(timezone.utc).isoformat(),
                "type": crumb_type,
                "category": category,
                "message": message,
                "level": level,
            })
            if len(self._breadcrumbs) > 50:
                self._breadcrumbs = self._breadcrumbs[-50:]

    def flush(self, timeout: float = 5.0) -> bool:
        """Wait for all queued events to be sent within *timeout* seconds.

        Returns True if the queue was fully drained, False if the timeout elapsed.
        """
        # queue.join() has no timeout — use a polling approach that respects timeout.
        deadline = time.monotonic() + timeout
        while not self._queue.empty():
            remaining = deadline - time.monotonic()
            if remaining <= 0:
                return False
            time.sleep(min(0.05, remaining))
        # One final join attempt on an effectively-empty queue to process
        # items that were in-flight when empty() was last checked.
        self._queue.join()
        return True

    def close(self, timeout: float = 5.0) -> bool:
        """Flush pending events and stop the background worker."""
        self._stop.set()
        return self.flush(timeout)

    # ── Framework integrations ────────────────────────────────────────────────

    def instrument_fastapi(self, app: Any) -> Any:
        """Add Errly error capture to a FastAPI app."""
        try:
            from starlette.middleware.base import BaseHTTPMiddleware
            from starlette.requests import Request

            errly_client = self

            class ErrlyMiddleware(BaseHTTPMiddleware):
                async def dispatch(self, request: Request, call_next):
                    try:
                        return await call_next(request)
                    except Exception as exc:
                        event = errly_client._new_event("error")
                        event["exception"] = {
                            "type": type(exc).__name__,
                            "value": str(exc),
                        }
                        event["stacktrace"] = _parse_traceback(exc.__traceback__)
                        event["request"] = {
                            "url": str(request.url),
                            "method": request.method,
                            "headers": {"user-agent": request.headers.get("user-agent", "")},
                        }
                        errly_client._enqueue(event)
                        raise

            app.add_middleware(ErrlyMiddleware)
        except ImportError:
            pass
        return app

    def instrument_flask(self, app: Any) -> Any:
        """Add Errly error capture to a Flask app."""
        errly_client = self

        @app.errorhandler(Exception)
        def handle_exception(exc: Exception):
            event = errly_client._new_event("error")
            event["exception"] = {"type": type(exc).__name__, "value": str(exc)}
            event["stacktrace"] = _parse_traceback(exc.__traceback__)
            errly_client._enqueue(event)
            raise exc

        return app

    # ── Internal ──────────────────────────────────────────────────────────────

    def _new_event(self, level: str) -> dict:
        with self._lock:
            crumbs = list(self._breadcrumbs)
            user = dict(self._user) if self._user else None

        return {
            "id": str(uuid.uuid4()).replace("-", ""),
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "level": level,
            "message": "",
            "platform": "python",
            "environment": self.environment,
            "release": self.release,
            "project_key": self.project,
            "breadcrumbs": crumbs,
            "user": user,
        }

    def _enqueue(self, event: dict) -> None:
        try:
            self._queue.put_nowait(event)
        except queue.Full:
            pass  # Drop if queue is full

    def _run(self) -> None:
        while not self._stop.is_set():
            try:
                event = self._queue.get(timeout=0.5)
                self._send(event)
                self._queue.task_done()
            except queue.Empty:
                continue
        # Drain remaining items after stop signal
        while True:
            try:
                event = self._queue.get_nowait()
                self._send(event)
                self._queue.task_done()
            except queue.Empty:
                break

    def _send(self, event: dict) -> None:
        try:
            data = json.dumps(event).encode("utf-8")
            req = urllib_request.Request(
                f"{self.url}/api/v1/events",
                data=data,
                headers={
                    "Content-Type": "application/json",
                    "X-Errly-Key": self.api_key,
                },
                method="POST",
            )
            with urllib_request.urlopen(req, timeout=5) as resp:
                resp.read()
        except (URLError, OSError):
            pass  # Network errors are silently dropped


# ── Helpers ───────────────────────────────────────────────────────────────────

def _parse_traceback(tb) -> list[dict]:
    """Parse a traceback into Errly stack frames."""
    if tb is None:
        return []

    frames = []
    for frame_info in traceback.extract_tb(tb):
        frames.append({
            "filename": frame_info.filename,
            "function": frame_info.name,
            "lineno": frame_info.lineno,
            "colno": 0,
            "in_app": _is_in_app(frame_info.filename),
        })
    return list(reversed(frames))  # Most recent first


def _is_in_app(filename: str) -> bool:
    return (
        "site-packages" not in filename
        and "lib/python" not in filename
        and "<" not in filename
    )

File: sdk/go/errly.go
// Package errly provides a lightweight error monitoring SDK for Go.
//
// Usage:
//
//	client := errly.New("http://localhost:5080", "your-api-key",
//	    errly.WithProject("my-service"),
//	    errly.WithEnvironment("production"),
//	    errly.WithRelease("v1.2.3"),
//	)
//	defer client.Flush()
//
//	// Capture an error
//	client.CaptureError(ctx, err)
//
//	// Use as HTTP middleware
//	http.ListenAndServe(":5080", client.Middleware(myHandler))
package errly

import (
	"bytes"
	"context"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"net/http"
	"runtime"
	"strings"
	"sync"
	"time"
)

// Client is the Errly error monitoring client.
type Client struct {
	endpoint    string
	apiKey      string
	project     string
	environment string
	release     string
	queue       chan *Event
	wg          sync.WaitGroup
	mu          sync.RWMutex
	breadcrumbs []Breadcrumb
	user        *UserInfo
	closeOnce   sync.Once
}

// Option configures the client.
type Option func(*Client)

func WithProject(project string) Option {
	return func(c *Client) { c.project = project }
}

func WithEnvironment(env string) Option {
	return func(c *Client) { c.environment = env }
}

func WithRelease(release string) Option {
	return func(c *Client) { c.release = release }
}

// New creates a new Errly client. Call Flush() or defer it to ensure all events are sent.
func New(endpoint, apiKey string, opts ...Option) *Client {
	c := &Client{
		endpoint:    strings.TrimRight(endpoint, "/"),
		apiKey:      apiKey,
		environment: "production",
		queue:       make(chan *Event, 512),
	}
	for _, o := range opts {
		o(c)
	}
	go c.worker()
	return c
}

// CaptureError captures an error with optional context keys.
func (c *Client) CaptureError(ctx context.Context, err error, extra ...map[string]any) string {
	if err == nil {
		return ""
	}
	event := c.newEvent("error")
	event.Exception = &Exception{
		Type:  errorType(err),
		Value: err.Error(),
	}
	event.Stacktrace = captureStack(2)
	if len(extra) > 0 {
		event.Extra = extra[0]
	}
	c.enqueue(event)
	return event.ID
}

// CaptureMessage captures a plain message at the given level.
func (c *Client) CaptureMessage(ctx context.Context, level, message string, extra ...map[string]any) string {
	event := c.newEvent(level)
	event.Message = message
	if len(extra) > 0 {
		event.Extra = extra[0]
	}
	c.enqueue(event)
	return event.ID
}

// SetUser sets the current user on future events.
func (c *Client) SetUser(user *UserInfo) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.user = user
}

// AddBreadcrumb adds a breadcrumb to the trail.
func (c *Client) AddBreadcrumb(category, message, level string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.breadcrumbs = append(c.breadcrumbs, Breadcrumb{
		Timestamp: time.Now().UTC(),
		Category:  category,
		Message:   message,
		Level:     level,
		Type:      "default",
	})
	if len(c.breadcrumbs) > 50 {
		c.breadcrumbs = c.breadcrumbs[1:]
	}
}

// Middleware wraps an http.Handler to automatically capture panics and errors.
func (c *Client) Middleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		defer func() {
			if rec := recover(); rec != nil {
				var err error
				switch v := rec.(type) {
				case error:
					err = v
				default:
					err = fmt.Errorf("%v", v)
				}
				event := c.newEvent("fatal")
				event.Exception = &Exception{Type: "panic", Value: err.Error()}
				event.Stacktrace = captureStack(3)
				event.Request = &RequestInfo{
					URL:    r.URL.String(),
					Method: r.Method,
					Headers: map[string]string{
						"User-Agent": r.Header.Get("User-Agent"),
					},
				}
				c.enqueue(event)
				http.Error(w, "Internal Server Error", http.StatusInternalServerError)
			}
		}()
		next.ServeHTTP(w, r)
	})
}

// Flush waits for all queued events to be sent (max 5s). Safe to call multiple times.
func (c *Client) Flush() {
	c.closeOnce.Do(func() { close(c.queue) })
	done := make(chan struct{})
	go func() {
		c.wg.Wait()
		close(done)
	}()
	select {
	case <-done:
	case <-time.After(5 * time.Second):
	}
}

// --- internals ---

func (c *Client) newEvent(level string) *Event {
	c.mu.RLock()
	crumbs := make([]Breadcrumb, len(c.breadcrumbs))
	copy(crumbs, c.breadcrumbs)
	user := c.user
	c.mu.RUnlock()

	return &Event{
		ID:          generateID(),
		Timestamp:   time.Now().UTC(),
		Level:       level,
		Platform:    "go",
		Environment: c.environment,
		Release:     c.release,
		ProjectKey:  c.project,
		Breadcrumbs: crumbs,
		User:        user,
	}
}

// enqueue adds an event to the send queue. wg.Add is called before enqueuing
// so that Flush()'s wg.Wait() never races with the worker's wg.Add.
func (c *Client) enqueue(event *Event) {
	c.wg.Add(1)
	select {
	case c.queue <- event:
	default:
		// Queue full — cancel the Add we just registered and drop the event.
		c.wg.Done()
	}
}

func (c *Client) worker() {
	httpClient := &http.Client{Timeout: 5 * time.Second}
	for event := range c.queue {
		c.send(httpClient, event)
		c.wg.Done()
	}
}

func (c *Client) send(httpClient *http.Client, event *Event) {
	data, err := json.Marshal(event)
	if err != nil {
		return
	}
	req, err := http.NewRequest("POST", c.endpoint+"/api/v1/events", bytes.NewReader(data))
	if err != nil {
		return
	}
	req.Header.Set("Content-Type", "application/json")
	if c.apiKey != "" {
		req.Header.Set("X-Errly-Key", c.apiKey)
	}
	httpClient.Do(req) //nolint:errcheck
}

func captureStack(skip int) []StackFrame {
	var frames []StackFrame
	for i := skip; i < skip+20; i++ {
		pc, file, line, ok := runtime.Caller(i)
		if !ok {
			break
		}
		fn := runtime.FuncForPC(pc)
		name := "unknown"
		if fn != nil {
			name = fn.Name()
		}
		frames = append(frames, StackFrame{
			Filename: file,
			Function: name,
			Lineno:   line,
			InApp:    !isSystemFrame(file),
		})
	}
	return frames
}

func isSystemFrame(file string) bool {
	return strings.Contains(file, "runtime/") ||
		strings.Contains(file, "net/http") ||
		strings.Contains(file, "go/src")
}

func errorType(err error) string {
	t := fmt.Sprintf("%T", err)
	if t == "*errors.errorString" || t == "*fmt.wrapError" {
		return "Error"
	}
	parts := strings.Split(t, ".")
	return parts[len(parts)-1]
}

func generateID() string {
	b := make([]byte, 16)
	rand.Read(b)
	return hex.EncodeToString(b)
}

// Re-export types so callers don't need to use internal package
type Event struct {
	ID          string            `json:"id"`
	Timestamp   time.Time         `json:"timestamp"`
	Level       string            `json:"level"`
	Message     string            `json:"message"`
	Exception   *Exception        `json:"exception,omitempty"`
	Stacktrace  []StackFrame      `json:"stacktrace,omitempty"`
	User        *UserInfo         `json:"user,omitempty"`
	Tags        map[string]string `json:"tags,omitempty"`
	Extra       map[string]any    `json:"extra,omitempty"`
	Breadcrumbs []Breadcrumb      `json:"breadcrumbs,omitempty"`
	Environment string            `json:"environment"`
	Release     string            `json:"release"`
	Platform    string            `json:"platform"`
	ProjectKey  string            `json:"project_key"`
	Fingerprint string            `json:"fingerprint"`
	Request     *RequestInfo      `json:"request,omitempty"`
}

type Exception struct {
	Type   string `json:"type"`
	Value  string `json:"value"`
	Module string `json:"module,omitempty"`
}

type StackFrame struct {
	Filename string `json:"filename"`
	Function string `json:"function"`
	Lineno   int    `json:"lineno"`
	Colno    int    `json:"colno,omitempty"`
	InApp    bool   `json:"in_app"`
}

type UserInfo struct {
	ID       string `json:"id,omitempty"`
	Email    string `json:"email,omitempty"`
	Username string `json:"username,omitempty"`
}

type Breadcrumb struct {
	Timestamp time.Time `json:"timestamp"`
	Type      string    `json:"type"`
	Category  string    `json:"category"`
	Message   string    `json:"message"`
	Level     string    `json:"level"`
}

type RequestInfo struct {
	URL     string            `json:"url,omitempty"`
	Method  string            `json:"method,omitempty"`
	Headers map[string]string `json:"headers,omitempty"`
}
